Support updates via UEFI capsules

SergiiDmytruk opened the pull request:
Originally: capsule parsing and coalescing.
Now: integration branch for UEFI capsules.

Other coreboot PRs merged into this one:
* https://github.com/Dasharo/coreboot/pull/527
* https://github.com/Dasharo/coreboot/pull/511
* https://github.com/Dasharo/coreboot/pull/562
* https://github.com/Dasharo/coreboot/pull/563

Related EDK PRs:
* https://github.com/Dasharo/edk2/pull/131
* https://github.com/Dasharo/edk2/pull/145
* https://github.com/Dasharo/edk2/pull/147
* https://github.com/Dasharo/edk2/pull/166
* https://github.com/Dasharo/edk2/pull/167

miczyg1 committed:
src/mainboard/msi: Enable dual TPM

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

SergiiDmytruk committed:
dasharo/uefi_capsules.c: coalesce and store UEFI capsules

How it approximately works:

(During a normal system run):
1. OS puts a capsule into RAM and calls UpdateCapsule() function of EFI
   runtime
2. If applying the update requires a reboot, EFI implementation creates
   a new CapsuleUpdateData* EFI variable pointing at the beginning of
   capsules description (not data, but description of the data) and does
   a warm reboot leaving capsule data and its description in RAM to be
   picked by firmware on the next boot process

(After DEV_INIT:)
3. Capsules are discovered by checking for CapsuleUpdateData* variables
4. Capsule description in memory and capsule data is validated for
   sanity
5. Capsule data is coalesced into a continuous piece of memory

(On BS_WRITE_TABLES via dasharo_add_capsules_to_bootmem() hook:)
6. Buffer with coalesced capsules is marked as reserved

(On BS_WRITE_TABLES via lb_uefi_capsules() hook:)
7. coreboot table entry is added for each of the discovered capsules

(In UEFI payload:)
8. CapsuleUpdateData* get removed
9. coreboot table is checked for any update capsules which are then
   applied

Change-Id: I162d678ae5c504906084b59c1a8d8c26dadb9433
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

macpijan commented on src/vendorcode/dasharo/uefi_capsules.c:
```diff
@@ -0,0 +1,769 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
```
@krystian-hebel @SergiiDmytruk Does it have to be dasharo-specific? Cannot it go into mainline coreboot?

SergiiDmytruk replied:
It doesn't.  I just didn't know that upstreaming was an explicit goal here.  Will move the file to `src/lib` if won't find any better place.

macpijan replied:
That was the main goal of the proposal: https://docs.dasharo.com/projects/capsule-updates/

We need to be in sync with the communities and establish a way to move forward since the very beginning, to ensure the whole process would go as smooth as possible.

SergiiDmytruk replied:
Made it more suitable for upstream: https://github.com/Dasharo/coreboot/compare/0ca09834457cc4c80472f049ad7d77d58b12f01a..1dbd00641ef16e4777f848694ec2c3131442e9a9

SergiiDmytruk commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,750 @@
+}
+
+/* Checks a single SG list for sanity.  Returns its end-of-list descriptor or
+ * an empty one on error. */
```
https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1bR292-R293

SergiiDmytruk commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,750 @@
+			goto error;
+		}
+
+		*total_data_size += ALIGN_UP(capsule_hdr->CapsuleImageSize, CAPSULE_ALIGNMENT);
```
Good point.

https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1bR305-R370

SergiiDmytruk commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,750 @@
+}
+
+/* Fills an array with pointers to capsule blocks.  Returns number of
+ * discovered capsule blocks or -1 on error. */
```
https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1bR377-R378

SergiiDmytruk commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,750 @@
+}
+
+/* Marks structures and data of SG lists as BM_MEM_RESERVED so we don't step on
+ * them when looking for usable memory. */
```
https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1bR474-R475

SergiiDmytruk commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,750 @@
+	struct block_descr seq_start = {0};
+
+	/* The code reserves sequences of blocks to avoid invoking
+	 * memranges_insert() on each on a bunch of adjacent 16-byte blocks. */
```
https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1bR484

SergiiDmytruk commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,750 @@
+	if (called)
+		return;
+
+	called = 1;
```
Dropped the flag.  I must have left it accidentally when trying to get rid of pre-processor at the bottom.

https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1bL634-L644

SergiiDmytruk commented on src/drivers/efi/Kconfig:
```diff
@@ -4,3 +4,11 @@ config DRIVERS_EFI_VARIABLE_STORE
+
+config DRIVERS_EFI_UPDATE_CAPSULES
+	bool "Include EFI update capsules driver"
+	depends on DRIVERS_EFI_VARIABLE_STORE
```
https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-7ffa6040ba95cc7406e87024010ef25159653dc594f56d00205c9fe95825f3ffR12

https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1bL646

SergiiDmytruk commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,750 @@
+		return;
+	}
+
+	init_pae_pagetables(&pae_page_tables);
```
https://github.com/Dasharo/coreboot/compare/196e3f62a00a5a6d989d6afeb57c825e5d0dac8c..c588ab4dfe61b5e6046afd58156de1fa9a77925e#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1bR644

SergiiDmytruk commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,750 @@
+
+	printk(BIOS_DEBUG, "capsules: chained capsule blocks.\n");
+
+	/* Reserve all blocks and the data they point to to avoid checking for
```
Not a typo, each "to" is necessary: "point to", "to avoid".

SergiiDmytruk committed:
configs/config.msi_*: enable UEFI capsule updates

Change-Id: I30058c4b4d43ef622bddfe69892d02ffffc039db
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented on configs/config.msi_ms7d25_ddr4:
```diff
@@ -7,25 +7,24 @@ CONFIG_IFD_BIN_PATH="3rdparty/dasharo-blobs/$(MAINBOARDDIR)/descriptor.bin"
-CONFIG_PCIEXP_DEFAULT_MAX_RESIZABLE_BAR_BITS=37
 CONFIG_PCIEXP_LANE_ERR_STAT_CLEAR=y
 CONFIG_DRIVERS_EFI_VARIABLE_STORE=y
-CONFIG_DRIVERS_EFI_UPDATE_CAPSULES=y
```
This suggests that capsules are enabled by default, I thought that no default meant disabled. I think this should be opt-in, at least for now.

SergiiDmytruk replied:
Good thing you've spotted it.  I just did it backwards: added option to defconfigs, re-generated them, then updated Kconfig to add the option.  No wonder it got removed.  Re-ordered commits and resaved configs.

SergiiDmytruk committed:
configs: bump EDK2 for UEFI capsules support

Change-Id: I02d917fb5f9c98fd2792e2c7d3cfaaf84b430ffd
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
drivers/efi: add optional ESRT-friendly coreboot table tag

New CONFIG_DRIVERS_EFI_FW_INFO is off by default, enabling it adds
DRIVERS_EFI_FW_{GUID,VERSION,LSV} to be used to specify firmware
version/update information.

Existing forms of versions wouldn't be sufficient because there is no
universal way of conversion to 32-bit unsigned integers and there are no
GUIDs or lowest supported versions.

Change-Id: Ic1b768d7bed43edf7ca8e41552087734054de033
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

miczyg1 commented on configs/config.msi_ms7d25_ddr4:
```diff
@@ -60,8 +61,4 @@ CONFIG_EDK2_DASHARO_MEMORY_CONFIG=y
 CONFIG_EDK2_RAM_DISK_ENABLE=y
-CONFIG_EDK2_CUSTOM_BUILD_PARAMS=""
 CONFIG_EDK2_CREATE_PREINSTALLED_BOOT_OPTIONS=y
-CONFIG_IPXE_ADD_SCRIPT=y
-CONFIG_IPXE_SCRIPT="3rdparty/dasharo-blobs/dasharo/dasharo.ipxe"
-CONFIG_IPXE_CUSTOM_BUILD_ID="0123456789"
```
The IPXE Kconfig options are broken. IIRC the name changed from `IPXE_*` to `PXE_*` and olddefconfig does not recognize these. It results in PXE being included as an external binary, instead of built from iPXE source. If you regenerate the defconfigs, play pay attention to how iPXE configuration looks like in menuconfig

SergiiDmytruk replied:
I should have checked it because this did look weird.  Upstream didn't rename those options, but updated `Kconfig` in a way that `CONFIG_BUILD_IPXE` now needs to be set explicitly.  https://github.com/Dasharo/coreboot/pull/509/commits/7a9776ca0d59889e3f63d565304a7942f15985a3

miczyg1 commented on src/drivers/efi/Kconfig:
```diff
@@ -6,3 +6,11 @@ config DRIVERS_EFI_VARIABLE_STORE
+
+config DRIVERS_EFI_UPDATE_CAPSULES
+	bool "Include EFI update capsules driver"
+	depends on DRIVERS_EFI_VARIABLE_STORE && SMMSTORE_V2
```
Isn't there any dependency on DRIVERS_EFI_FW_INFO from https://github.com/Dasharo/coreboot/pull/527 ?

I mean, capsules in EDK2 won't make sense if there is no ESRT/FMP, right?

SergiiDmytruk replied:
Rebased this PR on that one and updated.

https://github.com/Dasharo/coreboot/pull/509/commits/31467e195468dada6224dc7cd541d6f096a42494#diff-7ffa6040ba95cc7406e87024010ef25159653dc594f56d00205c9fe95825f3ff

krystian-hebel commented on src/commonlib/include/commonlib/coreboot_tables.h:
```diff
@@ -87,6 +87,7 @@ enum {
 	LB_TAG_TYPE_C_INFO		= 0x0042,
 	LB_TAG_ACPI_RSDP		= 0x0043,
 	LB_TAG_PCIE			= 0x0044,
+	LB_TAG_CAPSULE			= 0x0045,
 	LB_TAG_EFI_FW_INFO		= 0x0046,
```
Since these PRs have been reordered, it would make sense to change those tag numbers as well.

SergiiDmytruk replied:
Wanted to say that will will require changing the value in EDK and saw that it was `0xb0` there (used before making changes suitable for upstreaming) which needs to be changed anyway.  Updated: https://github.com/Dasharo/coreboot/compare/7a9776ca0d59889e3f63d565304a7942f15985a3..11de72c3ff2f755a0fb7ac45d38f93c5bf7a68c9#diff-8d6b99381f9485f5fbb941a0c702bad47ed392ba3058ede998847776409d37d2

krystian-hebel committed:
drivers/smmstore: add ability to write to whole flash

With DRIVERS_EFI_UPDATE_CAPSULES enabled, SMMSTORE SMI handler can use
commands with highest bit (0x80) set to access whole flash instead of
just the SMMSTORE region. The rest of interface is identical to regular
SMMSTORE v2.

Change-Id: I7f3dbfa965b9dcbade8b2f06a5bd2ac1345c7972
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,742 @@
+	enum { MAX_CAPSULE_BLOCKS = MAX_CAPSULES };
+
+	struct region_device rdev;
+	if (!smmstore_lookup_region(&rdev)) {
```
Something is wrong around here:

```
FMAP: area SMMSTORE found @ 0 (262144 bytes)
capsules: no SMMSTORE region, no update capsules.
```

SergiiDmytruk replied:
Inverted condition.  Must have messed it up while addressing review.  There was another issue as well.

https://github.com/Dasharo/coreboot/compare/11de72c3ff2f755a0fb7ac45d38f93c5bf7a68c9..628cce5b62a2df64b6150e4b63ef1fb715b1c6ea

krystian-hebel committed:
drivers/smmstore: add logic to disable capsule update handling code

This adds a call to SMMSTORE that saves information about availability
of capsules in SMM memory. This new call is ignored when run more than
once, which means that there should be no way of enabling full flash
handling after it was disabled and vice versa.

The call should be always made by the firmware to lock further calls,
otherwise OS could gain full flash access. This is done on entry to
BS_POST_DEVICE, after capsules are obtained in BS_DEV_INIT.

Change-Id: I3dc175ea313aae1edae304520595b82db7206cbb
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
Documentation/drivers/smmstorev2.md: describe capsule update API

This commit adds a description of capsule update commands implemented
in SMMSTORE in previous patches.

Change-Id: I94761d18be567e5302d1a836f09f0a7eecb4fb00
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

SergiiDmytruk commented:
Fixed GUID of `CapsuleUpdate*` variables and an issue with coalesced capsule overwriting cbmem.

krystian-hebel commented:
Same story as https://github.com/Dasharo/coreboot/pull/527, but since this has better branch name, perhaps closing #527 and using this as a target would make more sense. In any case, there are comments that hasn't been marked as resolved (@miczyg1 ping), I won't close the other PR until then to not mess up with how GH shows the changes.

miczyg1 commented on src/drivers/efi/capsules.c:
```diff
@@ -0,0 +1,759 @@
+ */
+struct block_descr {
+	uint64_t self;
+
```
```suggestion
```

miczyg1 replied:
Is it intentional?

SergiiDmytruk replied:
It is intentional to separate real fields from location.  Can add a comment to make it clearer.

miczyg1 replied:
That would help, otherwise it simply looks as a redundant newline

SergiiDmytruk replied:
https://github.com/Dasharo/coreboot/compare/0b4ea6501c3e8d88ccecb8c880b259c4b8949e43..a43a5cf6af820f6803bf3c6f35371c0c2a4b1d89#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1b

Also moved savedefconfig commit to https://github.com/Dasharo/coreboot/pull/519 and changed its base to `dasharo`.

miczyg1 replied:
Thank you.

SergiiDmytruk commented:
The last force push apart from updating EDK revision (and rebasing on `dasharo` just to not lag behind) extends firmware information to include firmware size.

SergiiDmytruk committed:
payloads/external/edk2: configure capsule updates

This requires version of EDK2 in use to understand those defines, but
the build isn't affected negatively if they aren't handled.  Upstream
EDK2 understands only CAPSULE_SUPPORT at the moment.

Change-Id: I1c684cb8929842a5d3c4b06e8a9c0a748470ea41
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk commented:
Last force push:
 - rebase on newer `dasharo`
 - update EDK hash to point to https://github.com/Dasharo/edk2/pull/154
 - add buffer allocation fix in https://github.com/Dasharo/coreboot/pull/509/commits/e4fcc9be50795b13e7c5836cbc69dcd07d528499 (to be squashed later)
 - update EDK-related part of build system to enable capsules

SergiiDmytruk commented:
Force-push: rebase to squash and update one of commit messages, also sync top-most commit with the [Gerrit](https://review.coreboot.org/q/topic:%22uefi-capsules%22).

SergiiDmytruk committed:
capsule.sh: add script for working with capsules

Change-Id: Ia72cff286f2dd4399d7874c1defe114ef8d95f33
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk commented:
Force-push:

* update EDK's commit hash
* switch define from CAPSULE_ENABLE to CAPSULE_SUPPORT because that's what UefiPayloadPkg is using (despite tianocore's wiki) and that's how changes were sent upstream

SergiiDmytruk commented:
Force-push:
 - rebase
 - update EDK2 revision (see https://github.com/Dasharo/edk2/pull/147#issuecomment-2336841445)

SergiiDmytruk commented:
Force-push:
 * rebase (picks up boot logo for Q35)
 * resolve conflicts
 * point EDK revision at https://github.com/Dasharo/edk2/pull/166 so CI artifacts can be used to test it

SergiiDmytruk committed:
soc/intel/fast_spi/mmap_boot.c: allow mapping whole flash

Add CONFIG_EXT_BIOS_FILL_UP to make BIOS region include everything
preceding it in the flash.

Change-Id: I885252a488bd35fc3afef571e6178642a059f883
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk commented:
Force-push:
 * rebase
 * update revision of EDK2
 * squash coalescing fixes
 * make whole flash accessible to Fast SPI

krystian-hebel commented on src/soc/intel/common/block/fast_spi/mmap_boot.c:
```diff
@@ -114,13 +114,24 @@ static void bios_mmap_init(void)
+	 * -------------------------------
+	 *
+	 * Same as above, but part of the flash preceding the BIOS region gets mapped into the
+	 * extended window as well (as much of it as there is room left).
```
When BIOS region is smaller than 16M, the rest of flash is still mapped into extended window. This either must be documented (with at least a note on diagram above the function that this is special case, as well as a comment here) or code changed to add `bios_start` to `bios_size` right after it is obtained, so it will be used for calculation of fixed window.

SergiiDmytruk replied:
New version: d1b1e450f383703746d1d28fdafb02b173fc2c0c

SergiiDmytruk committed:
configs/config.msi_ms7e06_*: add CONFIG_OPTION_BACKEND_NONE=y

UEFI board variants use UEFI variables directly instead of using them
through options API.  More importantly, be consistent with Z690-A.

Change-Id: I283b2198dbbb81baebf84d2eff33c0cd238b118d
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
configs/config.msi_*: v1.1.4-rc0 and v0.9.2-rc0

This version is necessary to enable testing capsule updates first
introduced for these boards.

Change-Id: I9d709c08cf771b5ddecdda936d7507c7f81016a5
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
configs/config.msi_*: v1.1.4-rc1 and v0.9.2-rc1

This is real release candidate for these releases.

Change-Id: I6b48ca82f6fc756b6032ef0695309db06328251f
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk commented:
Force-push:
 - rebase
 - squash recent commits
 - update EDK2 revision

krystian-hebel committed:
configs/config.emulation_qemu_x86_q35_uefi: enable capsules

Change-Id: Ic7dbdc044ea7cd799eb29800a8245e67c67b90d1
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
configs/config.msi_*: enable CPU configuration menu

Allows specifying how many performance/efficiency cores should run and
whether hyper-threading is enabled.

Change-Id: I540305af30be59652429f44d58f094aed549e9d1
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
.github: check capsule configuration in defconfigs

Change-Id: I0e309d79cf4d58e8f5d7468c74eb2281ea4f9857
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
soc/alderlake/Makefile.mk: use microcode from dasharo-blobs

ms7d25 and ms7e06 contain the same microcode.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
configs/config.msi_*: disable iPXE serial console

It results in duplicated output.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
.github/workflows/build.yml: build capsules for MSI

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
configs/config.msi_*: v1.1.4-rc2 and v0.9.2-rc2

One more release candidate.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
configs/config.msi_*: v1.1.4-rc3 and v0.9.2-rc3

Another release candidate.

Change-Id: I4b1d607ad874c824ab99e17e218c58efe4193984
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

macpijan committed:
configs: msi release v1.1.4 and v0.9.2

Change-Id: I489db304d6dc51a64fa3f15c8ccfe10ff08d302d
Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

SergiiDmytruk commented:
@macpijan, you didn't update `CONFIG_DRIVERS_EFI_MAIN_FW_VERSION` in configs, was that intentional?  The last byte should be changed to `0x80` for a release (so that release version is bigger than any RC version).

macpijan commented:
> @macpijan, you didn't update CONFIG_DRIVERS_EFI_MAIN_FW_VERSION in configs, was that intentional? The last byte should be changed to 0x80 for a release (so that release version is bigger than any RC version).

@SergiiDmytruk @krystian-hebel I was simply not aware of this being also needed right now.

If we need to fix it, we can still do it I guess.

SergiiDmytruk commented:
CI [complains about it](https://github.com/Dasharo/coreboot/actions/runs/11840836680/job/32995624248?pr=509):
```
config.msi_ms7d25_ddr4: CONFIG_LOCALVERSION doesn't match CONFIG_DRIVERS_EFI_MAIN_FW_VERSION
                        v1.1.4 != v1.1.4-rc3 (0x01010403)
config.msi_ms7d25_ddr5: CONFIG_LOCALVERSION doesn't match CONFIG_DRIVERS_EFI_MAIN_FW_VERSION
                        v1.1.4 != v1.1.4-rc3 (0x01010403)
config.msi_ms7e06_ddr4: CONFIG_LOCALVERSION doesn't match CONFIG_DRIVERS_EFI_MAIN_FW_VERSION
                        v0.9.2 != v0.9.2-rc3 (0x00090203)
config.msi_ms7e06_ddr5: CONFIG_LOCALVERSION doesn't match CONFIG_DRIVERS_EFI_MAIN_FW_VERSION
                        v0.9.2 != v0.9.2-rc3 (0x00090203)
```

We don't rely on it in code and capsule updates should be fine, so it's not a hard requirement if there is any cost to updating configs again.

krystian-hebel commented:
I've updated the configs as a new rebase was needed anyway. When CI builds pass ~~this PR can be merged~~ https://github.com/Dasharo/edk2/pull/147 will have to be merged before this one, and it may also need a rebase...

krystian-hebel commented:
After consulting with Beata, I've restored this PR to `msi_ms7d25_v1.1.4-rc3` with just a change to local version and `CONFIG_DRIVERS_EFI_MAIN_FW_VERSION`. The release tag is supposed to be put on it, however we would have to keep the branch (not safe, someone may remove it) or tag (safer) edk2 version that this PR points to. Otherwise, after rebasing https://github.com/Dasharo/edk2/pull/147 the hash will change, and it would be impossible to build this release. (This happened before, for Q35 v0.2.0, it can no longer be built from the release tag).

SergiiDmytruk commented:
> (This happened before, for Q35 v0.2.0, it can no longer be built from the release tag).

Are you sure, seems to build?  https://github.com/Dasharo/edk2/commit/11b26796145e29b2ceeb5a00b130f81246115159 is there and it's part of some branches/tags.

krystian-hebel commented:
> Are you sure, seems to build? Dasharo/edk2@11b2679 is there and it's part of some branches/tags.

Huh, I must've be mistaken...
